File: lib/security/credentialEnvelope.ts
import {
  constants,
  createDecipheriv,
  createHash,
  createPrivateKey,
  createPublicKey,
  generateKeyPairSync,
  privateDecrypt,
  type KeyObject,
} from "crypto";
import { getSecretValue } from "@/lib/security/secretSource";

const ENVELOPE_VERSION = "1";
const ENVELOPE_ALGORITHM = "RSA-OAEP-256+A256GCM";
const AES_KEY_BYTES = 32;
const AES_GCM_IV_BYTES = 12;
const AES_GCM_TAG_BYTES = 16;

type CredentialKeyMaterial = {
  keyId: string;
  privateKey: KeyObject;
  publicKeySpkiDer: Buffer;
};

type DecryptedCredentialPayload = {
  email: string;
  password: string;
};

let cachedKeyMaterial: CredentialKeyMaterial | null = null;
let warnedAboutEphemeralKey = false;

export type CredentialPublicKeyDescriptor = {
  version: string;
  algorithm: string;
  keyId: string;
  publicKey: string;
};

export type CredentialEnvelopeInput = {
  encryptedPayload: string;
  encryptedKey: string;
  encryptedIv: string;
  keyId?: string | null;
  version?: string | null;
};

export function getCredentialPublicKeyDescriptor(): CredentialPublicKeyDescriptor {
  const keyMaterial = getCredentialKeyMaterial();
  return {
    version: ENVELOPE_VERSION,
    algorithm: ENVELOPE_ALGORITHM,
    keyId: keyMaterial.keyId,
    publicKey: encodeBase64Url(keyMaterial.publicKeySpkiDer),
  };
}

export function decryptCredentialEnvelope(
  envelope: CredentialEnvelopeInput,
): DecryptedCredentialPayload {
  const keyMaterial = getCredentialKeyMaterial();

  if (envelope.version && envelope.version !== ENVELOPE_VERSION) {
    throw new Error("unsupported_envelope_version");
  }
  if (envelope.keyId && envelope.keyId !== keyMaterial.keyId) {
    throw new Error("unknown_envelope_key");
  }

  const encryptedPayload = decodeBase64Url(envelope.encryptedPayload);
  const encryptedKey = decodeBase64Url(envelope.encryptedKey);
  const encryptedIv = decodeBase64Url(envelope.encryptedIv);

  if (encryptedIv.length !== AES_GCM_IV_BYTES) {
    throw new Error("invalid_envelope_iv");
  }
  if (encryptedPayload.length <= AES_GCM_TAG_BYTES) {
    throw new Error("invalid_envelope_payload");
  }

  const symmetricKey = privateDecrypt(
    {
      key: keyMaterial.privateKey,
      padding: constants.RSA_PKCS1_OAEP_PADDING,
      oaepHash: "sha256",
    },
    encryptedKey,
  );

  if (symmetricKey.length !== AES_KEY_BYTES) {
    throw new Error("invalid_envelope_key");
  }

  const ciphertext = encryptedPayload.subarray(0, -AES_GCM_TAG_BYTES);
  const authTag = encryptedPayload.subarray(-AES_GCM_TAG_BYTES);

  const decipher = createDecipheriv("aes-256-gcm", symmetricKey, encryptedIv);
  decipher.setAuthTag(authTag);

  const plaintextBuffer = Buffer.concat([
    decipher.update(ciphertext),
    decipher.final(),
  ]);
  const parsed = JSON.parse(plaintextBuffer.toString("utf8")) as {
    email?: unknown;
    password?: unknown;
  };

  const email =
    typeof parsed.email === "string" ? parsed.email.trim().toLowerCase() : "";
  const password = typeof parsed.password === "string" ? parsed.password : "";

  if (!email || !password) {
    throw new Error("invalid_envelope_credentials");
  }

  return { email, password };
}

function getCredentialKeyMaterial(): CredentialKeyMaterial {
  if (cachedKeyMaterial) {
    return cachedKeyMaterial;
  }

  const configuredPrivateKey = getSecretValue({
    envVar: "AUTH_CREDENTIALS_PRIVATE_KEY",
    fileEnvVar: "AUTH_CREDENTIALS_PRIVATE_KEY_FILE",
  });

  if (configuredPrivateKey) {
    const normalizedKey = configuredPrivateKey.replace(/\\n/g, "\n");
    const privateKey = createPrivateKey(normalizedKey);
    const publicKeySpkiDer = createPublicKey(privateKey).export({
      type: "spki",
      format: "der",
    }) as Buffer;
    cachedKeyMaterial = {
      keyId: deriveKeyId(publicKeySpkiDer),
      privateKey,
      publicKeySpkiDer,
    };
    return cachedKeyMaterial;
  }

  const generated = generateKeyPairSync("rsa", {
    modulusLength: 2048,
    privateKeyEncoding: {
      type: "pkcs8",
      format: "pem",
    },
    publicKeyEncoding: {
      type: "spki",
      format: "der",
    },
  });

  const privateKey = createPrivateKey(generated.privateKey);
  const publicKeySpkiDer = Buffer.from(generated.publicKey);

  if (!warnedAboutEphemeralKey) {
    warnedAboutEphemeralKey = true;
    console.warn(
      "[security] auth_credentials_private_key_missing using ephemeral login envelope key; set AUTH_CREDENTIALS_PRIVATE_KEY for stable multi-instance deployments",
    );
  }

  cachedKeyMaterial = {
    keyId: deriveKeyId(publicKeySpkiDer),
    privateKey,
    publicKeySpkiDer,
  };
  return cachedKeyMaterial;
}

function deriveKeyId(publicKeySpkiDer: Buffer): string {
  return createHash("sha256")
    .update(publicKeySpkiDer)
    .digest("base64url")
    .slice(0, 24);
}

function decodeBase64Url(value: string): Buffer {
  return Buffer.from(value, "base64url");
}

function encodeBase64Url(value: Buffer): string {
  return value.toString("base64url");
}

File: tday-backend/src/main/kotlin/com/ohmz/tday/security/CredentialEnvelope.kt
package com.ohmz.tday.security

import com.ohmz.tday.config.AppConfig
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import org.slf4j.LoggerFactory
import java.math.BigInteger
import java.security.KeyFactory
import java.security.KeyPairGenerator
import java.security.MessageDigest
import java.security.PrivateKey
import java.security.interfaces.RSAPrivateKey
import java.security.spec.PKCS8EncodedKeySpec
import java.security.spec.RSAPublicKeySpec
import java.util.Base64
import java.security.spec.MGF1ParameterSpec
import javax.crypto.Cipher
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.OAEPParameterSpec
import javax.crypto.spec.PSource
import javax.crypto.spec.SecretKeySpec

@Serializable
data class CredentialPublicKeyDescriptor(
    val version: String,
    val algorithm: String,
    val keyId: String,
    val publicKey: String,
)

data class DecryptedCredentials(val email: String, val password: String)

data class CredentialEnvelopeInput(
    val encryptedPayload: String,
    val encryptedKey: String,
    val encryptedIv: String,
    val keyId: String? = null,
    val version: String? = null,
)

interface CredentialEnvelope {
    fun getPublicKeyDescriptor(): CredentialPublicKeyDescriptor
    fun decrypt(envelope: CredentialEnvelopeInput): DecryptedCredentials
}

class CredentialEnvelopeImpl(private val config: AppConfig) : CredentialEnvelope {
    private val logger = LoggerFactory.getLogger(CredentialEnvelopeImpl::class.java)
    private val envelopeVersion = "1"
    private val envelopeAlgorithm = "RSA-OAEP-256+A256GCM"
    private val aesKeyBytes = 32
    private val aesGcmIvBytes = 12
    private val aesGcmTagBits = 128

    private data class KeyMaterial(
        val keyId: String,
        val privateKey: PrivateKey,
        val publicKeySpkiDer: ByteArray,
    )

    private val keyMaterial: KeyMaterial by lazy { loadKeyMaterial() }

    override fun getPublicKeyDescriptor(): CredentialPublicKeyDescriptor {
        val km = keyMaterial
        return CredentialPublicKeyDescriptor(
            version = envelopeVersion,
            algorithm = envelopeAlgorithm,
            keyId = km.keyId,
            publicKey = km.publicKeySpkiDer.toBase64Url(),
        )
    }

    override fun decrypt(envelope: CredentialEnvelopeInput): DecryptedCredentials {
        val km = keyMaterial
        if (envelope.version != null && envelope.version != envelopeVersion) {
            throw IllegalArgumentException("unsupported_envelope_version")
        }
        if (envelope.keyId != null && envelope.keyId != km.keyId) {
            throw IllegalArgumentException("unknown_envelope_key")
        }

        val encryptedPayload = envelope.encryptedPayload.fromBase64Url()
        val encryptedKey = envelope.encryptedKey.fromBase64Url()
        val encryptedIv = envelope.encryptedIv.fromBase64Url()

        if (encryptedIv.size != aesGcmIvBytes) throw IllegalArgumentException("invalid_envelope_iv")
        if (encryptedPayload.size <= 16) throw IllegalArgumentException("invalid_envelope_payload")

        val rsaCipher = Cipher.getInstance("RSA/ECB/OAEPPadding")
        val oaepParams = OAEPParameterSpec("SHA-256", "MGF1", MGF1ParameterSpec.SHA256, PSource.PSpecified.DEFAULT)
        rsaCipher.init(Cipher.DECRYPT_MODE, km.privateKey, oaepParams)
        val symmetricKey = rsaCipher.doFinal(encryptedKey)

        if (symmetricKey.size != aesKeyBytes) throw IllegalArgumentException("invalid_envelope_key")

        val aesCipher = Cipher.getInstance("AES/GCM/NoPadding")
        aesCipher.init(
            Cipher.DECRYPT_MODE,
            SecretKeySpec(symmetricKey, "AES"),
            GCMParameterSpec(aesGcmTagBits, encryptedIv),
        )

        val plaintext = String(aesCipher.doFinal(encryptedPayload), Charsets.UTF_8)
        val parsed = Json.decodeFromString<CredentialPayloadJson>(plaintext)

        val email = parsed.email?.trim()?.lowercase() ?: ""
        val password = parsed.password ?: ""
        if (email.isEmpty() || password.isEmpty()) throw IllegalArgumentException("invalid_envelope_credentials")

        return DecryptedCredentials(email, password)
    }

    @Serializable
    private data class CredentialPayloadJson(val email: String? = null, val password: String? = null)

    private fun loadKeyMaterial(): KeyMaterial {
        val configuredPem = config.credentialsPrivateKeyPem
        if (!configuredPem.isNullOrBlank()) {
            val normalizedPem = configuredPem.replace("\\n", "\n")
            val keyFactory = KeyFactory.getInstance("RSA")
            val pemBody = normalizedPem
                .replace("-----BEGIN PRIVATE KEY-----", "")
                .replace("-----END PRIVATE KEY-----", "")
                .replace("\\s".toRegex(), "")
            val keySpec = PKCS8EncodedKeySpec(Base64.getDecoder().decode(pemBody))
            val privateKey = keyFactory.generatePrivate(keySpec) as RSAPrivateKey

            val publicKey = KeyFactory.getInstance("RSA").generatePublic(
                RSAPublicKeySpec(privateKey.modulus, BigInteger.valueOf(65537)),
            )
            val publicKeyDer = publicKey.encoded

            return KeyMaterial(
                keyId = deriveKeyId(publicKeyDer),
                privateKey = privateKey,
                publicKeySpkiDer = publicKeyDer,
            )
        }

        logger.warn("[security] auth_credentials_private_key_missing using ephemeral login envelope key")
        val keyPair = KeyPairGenerator.getInstance("RSA").apply { initialize(2048) }.generateKeyPair()
        val publicKeyDer = keyPair.public.encoded

        return KeyMaterial(
            keyId = deriveKeyId(publicKeyDer),
            privateKey = keyPair.private,
            publicKeySpkiDer = publicKeyDer,
        )
    }

    private fun deriveKeyId(publicKeyDer: ByteArray): String {
        val digest = MessageDigest.getInstance("SHA-256").digest(publicKeyDer)
        return digest.toBase64Url().take(24)
    }
}
